relay-filter/src/generic.rs

```diff
@@ -27,32 +33,151 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+/// iterator yields filters according to the principles below:
+/// - Filters from project configs are evaluated before filters from global
+/// configs.
+/// - No duplicates: once a filter is evaluated (yielded or skipped), no filter
+/// with the same id is evaluated again.
+/// - If a filter with the same id exists in projects and global configs, both
+/// are merged and the filter is yielded. Values from the filter in the project
+/// config are prioritized.
+fn merge_generic_filters<'a>(
+    project: &'a GenericFiltersConfig,
+    global: Option<&'a GenericFiltersConfig>,
+    #[cfg(test)] version: u16,
+) -> impl Iterator<Item = GenericFilterConfigRef<'a>> {
+    #[cfg(not(test))]
+    let version = VERSION;
+
+    let is_supported =
+        project.version <= version && global.map_or(true, |gf| gf.version <= version);
+
+    is_supported
+        .then(|| {
+            DynamicGenericFiltersConfigIter::new(&project.filters, global.map(|gc| &gc.filters))
+        })
+        .into_iter()
+        .flatten()
+}
+
+/// Iterator over the generic filters of the project and global configs.
+struct DynamicGenericFiltersConfigIter<'a> {
+    /// Generic project filters.
+    project: &'a GenericFiltersMap,
+    /// Index of the next filter in project configs to evaluate.
+    project_index: usize,
+    /// Generic global filters.
+    global: Option<&'a GenericFiltersMap>,
+    /// Index of the next filter in global configs to evaluate.
+    global_index: usize,
+}
+
+impl<'a> DynamicGenericFiltersConfigIter<'a> {
+    pub fn new(project: &'a GenericFiltersMap, global: Option<&'a GenericFiltersMap>) -> Self {
+        DynamicGenericFiltersConfigIter {
+            project,
+            project_index: 0,
+            global,
+            global_index: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for DynamicGenericFiltersConfigIter<'a> {
+    type Item = GenericFilterConfigRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((id, filter)) = self.project.get_index(self.project_index) {
+            self.project_index += 1;
+            let merged = merge_filters(filter, self.global.and_then(|gf| gf.get(id)));
+            return Some(merged);
+        }
+
+        loop {
+            let (id, filter) = self.global?.get_index(self.global_index)?;
+            self.global_index += 1;
+            if !self.project.contains_key(id) {
+                return Some(filter.into());
+            }
+        }
+    }
+}
+
+impl<'a> FusedIterator for DynamicGenericFiltersConfigIter<'a> {}
+
+/// Merges the two filters with the same id, prioritizing values from the primary.
+///
+/// It's assumed both filters share the same id. The returned filter will have
+/// the primary filter's ID.
+fn merge_filters<'a>(
+    primary: &'a GenericFilterConfig,
+    secondary: Option<&'a GenericFilterConfig>,
+) -> GenericFilterConfigRef<'a> {
+    GenericFilterConfigRef {
+        id: primary.id.as_str(),
+        is_enabled: primary.is_enabled,
```

Comment:
If a filter is enabled by default in the global config, it's still possible to override it from the project config:
- Project config filters are evaluated first, along with the filters in the global configs, and have priority in this merging step.
- When going through the global config filters, those filters that exist in the project filters aren't evaluated again.
Let me know if there's a way to make this more clear.

Reply:
You're right, we don't need `is_enabled` to be `Option<bool>` because we can just omit the project-side config if we want the global default to work, and add a project-side config if we want to override. Sorry for the confusion.